## Problem 3
### (Problem 5, Chapter 5, page 198)
In Chapter 4, we used logistic regression to predict the probability of
default using income and balance on the Default data set. We will
now estimate the test error of this logistic regression model using the
validation set approach. Do not forget to set a random seed before
beginning your analysis.

In [85]:
#libraries
import os
import pandas as pd
import numpy as np
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import LabelBinarizer

In [86]:
#os.getcwd()
## loading the dataset
filePath = "/home/humble-fool/school_work/Spring_18/Data-Science-Lab-ee379k/datasets/ISLR/data/csv/"
fileName = "Default.csv"
default = pd.read_csv(filePath+fileName)
#default.head()

#### Part A

Fit a logistic regression model that uses income and balance to
predict default.

In [87]:
#splitting data into training and testing
clf = LogisticRegression()
X_train = default[['income','balance']]
y_train = default['default']
_= clf.fit(X_train, y_train)
pred_a = clf.predict(X_train)


In [88]:
#calculating training error
size_a = pred_a.size
count_a = 0
for i in range(size_a):
    if(pred_a[i]!=y_train.iloc[i]):
        count_a +=1
train_error = count_a/size_a
print("Training error of the dataset is {}".format(train_error))

Training error of the dataset is 0.0336


#### Part B

Using the validation set approach, estimate the test error of this
model. In order to do this, you must perform the following steps:
i. Split the sample set into a training set and a validation set.

ii. Fit a multiple logistic regression model using only the train-
ing observations.

iii. Obtain a prediction of default status for each individual in
the validation set by computing the posterior probability of
default for that individual, and classifying the individual to
the default category if the posterior probability is greater
than 0.5.

iv. Compute the validation set error, which is the fraction of
the observations in the validation set that are misclassified.

In [89]:
#Splitting data into test and train
train,test = train_test_split(default, test_size=0.2)
X_train_b = train[['income','balance']]
y_train_b = train['default']
X_test_b = test[['income','balance']]
y_test_b = test['default']

In [90]:
# fitting the logistic regression model and saving only concerned probabilities
clf_b = LogisticRegression()
_ = clf_b.fit(X_train_b,y_train_b)
pred_b = clf_b.predict(X_test_b)

In [91]:
#calculating the number of values that don't match across validation set and predictions
size_b = pred_b.size
count_b = 0
for i in range(size):
    if(pred_b[i]!=y_test_b.iloc[i]):
        count_b +=1
        
# calculating validation error
validation_error_b = count_b/y_test_b.size
print("test error with validation set ratio 0.2 is {}".format(validation_error_b))

test error with validation set ratio 0.2 is 0.0365


#### Part C
Repeat the process in (b) three times, using three different splits
of the observations into a training set and a validation set. Com-
ment on the results obtained.



In [92]:
# repeating part b with validation set of ratio 0.3
train_c_1,test_c_1 = train_test_split(default, test_size=0.3)
X_train_c_1 = train_c_1[['income','balance']]
y_train_c_1 = train_c_1['default']
X_test_c_1 = test_c_1[['income','balance']]
y_test_c_1 = test_c_1['default']


# fitting the logistic regression model and saving only concerned probabilities
clf_c_1 = LogisticRegression()
_ = clf_c_1.fit(X_train_c_1,y_train_c_1)
pred_c_1 = clf_c_1.predict(X_test_c_1)

#calculating the number of values that don't match across validation set and predictions
size_c_1 = pred_c_1.size
count_c_1 = 0
for i in range(size_c_1):
    if(pred_c_1[i]!=y_test_c_1.iloc[i]):
        count_c_1 +=1
        
# calculating validation error
validation_error_c_1 = count_c_1/y_test_c_1.size
print("Test error with validation set ratio 0.3 is {}".format(validation_error_c_1))



Test error with validation set ratio 0.3 is 0.035


In [93]:
# repeating part b with validation set of ratio 0.4
train_c_2,test_c_2 = train_test_split(default, test_size=0.4)
X_train_c_2 = train_c_2[['income','balance']]
y_train_c_2 = train_c_2['default']
X_test_c_2 = test_c_2[['income','balance']]
y_test_c_2 = test_c_2['default']


# fitting the logistic regression model and saving only concerned probabilities
clf_c_2 = LogisticRegression()
_ = clf_c_2.fit(X_train_c_2,y_train_c_2)
pred_c_2 = clf_c_2.predict(X_test_c_2)

#calculating the number of values that don't match across validation set and predictions
size_c_2 = pred_c_2.size
count_c_2 = 0
for i in range(size_c_2):
    if(pred_c_2[i]!=y_test_c_2.iloc[i]):
        count_c_2 +=1
        
# calculating validation error
validation_error_c_2 = count_c_2/y_test_c_2.size
print("Test error with validation set ratio 0.4 is {}".format(validation_error_c_2))

Test error with validation set ratio 0.4 is 0.03


In [94]:
# repeating part b with validation set of ratio 0.5
train_c_3,test_c_3 = train_test_split(default, test_size=0.5)
X_train_c_3 = train_c_3[['income','balance']]
y_train_c_3 = train_c_3['default']
X_test_c_3 = test_c_3[['income','balance']]
y_test_c_3 = test_c_3['default']


# fitting the logistic regression model and saving only concerned probabilities
clf_c_3 = LogisticRegression()
_ = clf_c_3.fit(X_train_c_3,y_train_c_3)
pred_c_3 = clf_c_3.predict(X_test_c_3)

#calculating the number of values that don't match across validation set and predictions
size_c_3 = pred_c_3.size
count_c_3 = 0
for i in range(size_c_3):
    if(pred_c_3[i]!=y_test_c_3.iloc[i]):
        count_c_3 +=1
        
# calculating validation error
validation_error_c_3 = count_c_3/y_test_c_3.size
print("Test error with validation set ratio 0.5 is {}".format(validation_error_c_3))

Test error with validation set ratio 0.5 is 0.035


Test error increases with increase in size of validation set as expected except for the case of validation set ratio 0.5

#### Part D

Now consider a logistic regression model that predicts the prob-
ability of default using income , balance , and a dummy variable
for student . Estimate the test error for this model using the val-
idation set approach. Comment on whether or not including a
dummy variable for student leads to a reduction in the test error
rate.

In [95]:
## converting student to float because regression model doesn't accept strings
default_copy = default.copy()
lb = LabelBinarizer()
y = lb.fit_transform(default.student)
default_copy['student'] = y

In [96]:
# considering a logistic regression model with student as a predictor 

train_d,test_d = train_test_split(default_copy, test_size=0.4)

X_train_d = train_d[['student','income','balance']]
y_train_d = train_d['default']
X_test_d = test_d[['student','income','balance']]
y_test_d = test_d['default']


# fitting the logistic regression model and saving only concerned probabilities
clf_d = LogisticRegression()
_ = clf_d.fit(X_train_d,y_train_d)
pred_d = clf_d.predict(X_test_d)

#calculating the number of values that don't match across validation set and predictions
size_d = pred_d.size
count_d = 0
for i in range(size_d):
    if(pred_d[i]!=y_test_d.iloc[i]):
        count_d +=1
        
# calculating validation error
validation_error_d = count_d/y_test_d.size
print("Test error with validation set ratio 0.4 is {}".format(validation_error_d))

Test error with validation set ratio 0.4 is 0.03475


Considering student status increases test error of the logistic regression model slightly as we observed change from .03425 to .03475 with same size validation set.